themes/default/content/blog/troubleshooting-builds-with-diagnostics/index.md

```diff
@@ -0,0 +1,154 @@
+I0328 17:44:36.319427   37108 backend.go:409] found username for access token
+Previewing update (dev)
+...
+```
+
+We can walk through this, though! Like most logging data, there's some initial labels, timestamps, and process data in the beginning of the line. But the part you want to focus on is the part just before that closing bracket (`]`). In this snippet, you can find the calls to various parts of the Pulumi deployment engine, like `backend.go` or `api.go` (and their relevant lines in the code, so [backend.go:409](https://github.com/pulumi/pulumi/blob/master/pkg/backend/httpstate/backend.go#L409) for that first line). The human-readable messages do give some context, as well. You can use this data to understand where we are in the calls that the engine makes. There's also plugins, which are the various providers. You might find `langruntime_plugin.go`, which reports on what the language host is doing, or `eventsink.go`, which shares which events (or actions happening in the system) are occurring when. As plugins are getting installed, you likely will find a bunch of "skipping file in plugin directory" messages if you've been running Pulumi for a bit. That's all of the older versions of plugins in your Pulumi plugin directory being checked and skipped if a newer one is available and required. Eventually, you'll get to a bunch of [gRPC](https://grpc.io/) calls, which are [remote procedure calls](https://en.wikipedia.org/wiki/Remote_procedure_call) that basically provide connecting glue among all of the various subroutines that are running on both the local machine and the remote Pulumi Service.
+
+## Fixing errors
+
+Sometimes, when exploring the diagnostic data, you can find out what's up pretty quickly by checking the last 10 or 20 lines. You always want to check a handful of lines backwards when going through this kind of information because the problem spots will often appear before child processes sending information and shutting down, other processes failing due to dependencies, etc. I often scan a bit up to something that appears familiar, such as a handful of lines I can find in a previous run, and then work my way down from there until I identify a problem spot.
+
+If it's not a problem in your code, however, it can be difficult to track down the error in question since there are so many moving parts. Based on the behavior of the program before it failed, you can track down which part you need to search for. If a provider is failing, for example, I'd check to find out whether the right versions of provider plugins are being installed, which requires searching the data for `plugins.go` or another similar process. If I'm running into an authentication problem, I'll check the `api.go` details and search for anything out of place.
```

Comment:
This “if it was a provider I’d check for what version” kind of thing is what I think could make this blog really useful. Walk through a handful of problems and how to find the answer from the logs.